### ISSUE TITLE
Refactor and add support for better error log message testing

### ISSUE THREAD

**Issue by DanRStevens** [OutpostUniverse/op2ext#131]
This might be a good problem to try Google Mock on. If we can create a mock object for the logger, than we can verify more directly if we are logging proper error messages, without needing to actually go to disk.

To test with mock objects (with or without Google Mock), we'd need to make a few changes to the logging code. We'd have to split the `Logger` class into an interface class, and a separate implementation. This might be a good idea to do anyway, since it would allow for multiple logging output options.

We would also need to change the global `logger` variable to be a pointer or reference to a `Logger`. That would allow method calls to go through to a potential derived class. I'd recommend a pointer if we want to allow for updates to the logging output at runtime, allowing it to be swapped out for a different type of logger. A global reference can only be set once, and it would be done during global initialization.

Not required, but something I was thinking about. Perhaps we should use global logging functions, rather than a global logging object. The global functions can then delegate to the logging object internally. That would allow restricting access to the logging object, and possibly making it not so global. That could have benefits for deferring the creation of the logging object. It would also protect the code from having to change `logger.Log` to `logger->Log` all over the place if we change it from a `Logger` instance to a pointer.

Finally, with Google Mock / Google Test support, we might potentially use the string matchers on the error messages. That would allow for testing of keywords, rather than exact error messages. Small changes in the error messages wouldn't immediately break tests, and we also wouldn't have to export the error strings to give tests access. As an example, suppose a file API has an error accessing a file. We might check that the error message contains the filename. Or we might check that the error message contains the path and filename if the directory portion is likely to be relevant.

As a final thought, if we want to allow for parallel unit tests on multiple threads, the (global) `logger` variable/pointer could be made into a thread local variable. That way each unit test, running on it's own thread, could have it's own mock logger object, allowing the unit tests to run in parallel without stepping on each other. This might be a good idea anyway, as logging from multiple threads would either need separate loggers, or the logger would have to be multithread aware and provide for proper safety or synchronization.

**Comment by Brett208** [comment]
I think you are discussing using the Singleton pattern for our logger (https://en.wikipedia.org/wiki/Singleton_pattern). I think it is an acceptable fit.

You also mentioned switching to using an external logger. I'd be happy to try that out as well. If we define our own interface, we could use an external logger without letting it be tightly coupled to our own code base. Maybe best to write the interface first. 

This is all a lot of work though, and maybe better on the backburner for now? Maybe what I'm getting at is that I don't plan to make this a priority but would welcome the changes.

-Brett

**Comment by DanRStevens** [comment]
Yes, basically a Singleton. I plan to hide away the objects calls within global functions. I guess technically that makes it a bit inside out from the typical singleton pattern. You can still instantiate multiple loggers of any type if you wish, perhaps for some custom logging of one section of code. It's just that there will only ever be one globally installed logger.

----

I noticed something a bit funny in our logging code. I think we got a bit lucky actually. I suspect it could technically result in undefined behaviour, though the way it's written, nothing bad is likely to happen.

In the `Logger` constructor, if it fails to open the file, it calls `PostErrorMessage`. In `PostErrorMessage`, just before the modal box is displayed, it calls `logger.Log`. There's a bit of a circular dependency here.

Now, once the `Logger` constructor starts, all subobjects have been constructed, namely the `logFile` object. That means the constructor is free to use any of its own objects. It's also free to call its own methods, provided they were designed to be safe to call from the constructor before it has finished. That is, it's free to call methods that have no precondition on the object state which the constructor had yet to setup.

The `Logger` instance itself however, is not yet fully constructed, as we are still in the constructor. That means outside calls to member functions produce undefined behaviour. In particular, they might call a method that has a precondition on the object state, which the constructor has yet to setup. The reference from `PostErrorMessage` is an outside call, and so shouldn't be allowed yet until the `Logger` constructor finishes.

The `Logger` class methods only really have one precondition on their use, in that they need the `logFile` object to be in a valid state. Fortunately the `logFile` object has already been initialized before entering the constructor. As the state has already been set to something valid, nothing bad happens.

So yeah, a read-only log file may trigger some slight undefined behaviour that we are getting away with.

**Comment by DanRStevens** [comment]
I did a bit of reading on Google Mock, and tried creating a mock `Logger` class for use with testing. The documentation looked to be straightforward, but it seems there's a version mismatch between the documentation I was reading, and the version of Google Mock I had installed.

Although my initial attempts didn't work, here's approximately what a Google Mock test might look like:
```
class MockLogger : public Logger {
public:
	MOCK_METHOD(void, Log, (const std::string& message, const std::string& moduleName), (override));
};

TEST(Log, SetLoggerReceivesMessages) {
	const auto message = std::string("Logger should receive this");

	MockLogger logger;
	EXPECT_CALL(logger, Log(message));

	EXPECT_NO_THROW(SetLogger(&logger));
	EXPECT_NO_THROW(Log(message));
	EXPECT_NO_THROW(SetLogger(nullptr));
}
```

**Comment by DanRStevens** [comment]
I think we should close out this issue.

We've restructured the logging code as suggested in the lead post.

The Google Mock test code was eventually written, and has already been merged. We just had to use an older version of the Google Mock macros to mock methods, and it worked.

The threading idea is perhaps out of scope. To support it properly you could derive new Logger classes with synchronization support. Such a solution would be an external concern to the rest of the logging code. We certainly don't need support for it now.

Writing actual tests for existing logging code is perhaps more the focus of Issue #107. Perhaps we can do a bit of issue renaming and continue the remaining work there.
